### PR TITLE
prevent user role lock

### PR DESCRIPTION
**Pull request by JanKaczmarkiewicz**
this implements https://github.com/gravitational/teleport/pull/15706 RFD approach

This enforces the state when at least one user has permission to edit roles.

This state can be damaged by (and approach to validate them):
| operation | check |
| - | - |
| deleting user | check if there is another user with permission to edit roles |
| editing user | if the user loses permission to edit roles we need to check if there is another user with permission to edit roles  |
| upserting user | see editing user check |
| deleting role | This does not require validation since deleting roles in use is not allowed. |
| editing role | if the role loses permission to edit roles we need to check if there will be some user with permission to edit roles after operation |
| upserting role | see editing role check |

### Changes
| description  | result |
| ------------- | ------------- |
| removing the last local user with permission to edit roles |  Screenshot 2022-09-06 at 17 07 33  |
| removing permission to edit roles from the last local user that is capable | Screenshot 2022-09-06 at 17 09 53  |
| removing permission to edit roles from the only role that is allowing them | Screenshot 2022-09-06 at 17 14 49  |
| removing role the only role that allows to edit | Screenshot 2022-09-06 at 17 18 38  |

This will work for both custom and default roles. The same error messages will be visible when running actions via `tctl`

### CODE REVIEW

**Comment by russjones** [comment]
@codingllama @nklaassen Can you review this PR and RFD as well? https://github.com/gravitational/teleport/pull/15706

File: lib/auth/access.go [hunk 1]
```diff
@@ -144,3 +148,71 @@ func (a *Server) DeleteLock(ctx context.Context, lockName string) error {
...
+		if !rule.HasResource(types.KindRole) {
+			return false
+		}
+
+		if !rule.HasVerb(types.VerbUpdate) {
+			return false
+		}
+
+		return true
+	})
```
Comment (nklaassen) [inline lib/auth/access.go:217]: The FP is fun, but I think boring `for` loops are the idiomatic way to do this in Go (and they're somewhat more efficient)

File: lib/auth/access.go [hunk 2]
```diff
@@ -144,3 +148,71 @@ func (a *Server) DeleteLock(ctx context.Context, lockName string) error {
...
 	return nil
 }
+
+// checkRoleRulesConstraint checks if the request will result in having
+// no roles with rules to upsert roles.
+func (s *Server) checkRoleRulesConstraint(ctx context.Context, targetRole types.Role, request string) error {
+	rolesWithUpdateRolesRule, err := s.getRolesWithUpdateRolesRule(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
```
Comment (nklaassen) [inline lib/auth/access.go:158]: Let's avoid fetching/filtering all of the roles in the cluster until after we check that this edit actually removes an update roles rule

File: lib/auth/access.go [hunk 3]
```diff
@@ -144,3 +148,71 @@ func (a *Server) DeleteLock(ctx context.Context, lockName string) error {
...
+		rolesWithoutTarget := Filter(u.GetRoles(), func(role string) bool {
+			return role != targetRole.GetName()
+		})
+
+		return isLocalUser && Some(rolesWithoutTarget, func(role string) bool {
+			return Contains(rolesWithUpdateRolesRule, role)
+		})
+	})
+
+	if len(localUsersWithPermissionToEditRoles) > 0 {
```
Comment (nklaassen) [inline lib/auth/access.go:183]: are we not going to support having a cluster with zero local users? was there a product decision on that?

File: lib/auth/fp.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
+package auth
```
Comment (nklaassen) [inline lib/auth/fp.go:1]: I'm not sure the auth package is the right place for these but I hesitate to suggest dumping them into `utils`, perhaps a new package?

at a minimum you'll need a license header here

File: lib/auth/user.go [hunk 5]
```diff
@@ -209,11 +224,54 @@ func (s *Server) DeleteUser(ctx context.Context, user string) error {
...
+func (s *Server) checkUserRoleConstraint(ctx context.Context, user types.User, request string) error {
+	rolesWithUpdateRolesRule, err := s.getRolesWithUpdateRolesRule(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	allUsers, err := s.Identity.GetUsers(false)
+	if err != nil {
+		return trace.Wrap(err)
+	}
```
Comment (nklaassen) [inline lib/auth/user.go:247]: IMO let's avoid fetching all roles and users in the cluster until we at least check that the user is losing a role

File: lib/auth/user.go [hunk 6]
```diff
@@ -209,11 +224,54 @@ func (s *Server) DeleteUser(ctx context.Context, user string) error {
...
+
+	isUserWithPermissionToEditRoles := Some(localUsersWithPermissionToEditRoles, func(u types.User) bool {
+		return user.GetName() == u.GetName()
+	})
+
+	isUpdatedUserWithPermissionToEditRoles := Some(user.GetRoles(), func(role string) bool {
+		return Contains(rolesWithUpdateRolesRule, role)
+	})
+
+	isUserLoosingPermissionToEditRoles := isUserWithPermissionToEditRoles && !isUpdatedUserWithPermissionToEditRoles
```
Comment (nklaassen) [inline lib/auth/user.go:269]:
```suggestion
	isUserLosingPermissionToEditRoles := isUserWithPermissionToEditRoles && !isUpdatedUserWithPermissionToEditRoles
```

**Comment by codingllama** [comment]
> @codingllama @nklaassen Can you review this PR and RFD as well? #15706

@russjones 1st RFD pass done.

@JanKaczmarkiewicz, is this PR predicated on RFD approval? Are you seeking approval from Nic and/or myself, or is our review more of an "informal" double-check? (Just so we know how to sort/prioritize things out.)

File: lib/auth/user.go (hunk 5 above)
Comment (JanKaczmarkiewicz) [reply lib/auth/user.go]: See updated code

File: lib/auth/access.go (hunk 2 above)
Comment (JanKaczmarkiewicz) [reply lib/auth/access.go]: Updated

File: lib/auth/access.go (hunk 1 above)
Comment (JanKaczmarkiewicz) [reply lib/auth/access.go]: I have removed fp module: https://github.com/gravitational/teleport/commit/d4e2fb2c21094c1de433e01633a104cc616e9467

**Pull request closed without merging**